runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.core;
+
+import java.util.Map;
+import org.apache.beam.runners.core.ExecutionContext.StepContext;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.transforms.Aggregator;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Sum;
+import org.apache.beam.sdk.transforms.reflect.DoFnSignature;
+import org.apache.beam.sdk.transforms.reflect.DoFnSignatures;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.WindowTracing;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.util.state.State;
+import org.apache.beam.sdk.util.state.StateSpec;
+import org.joda.time.Instant;
+
+/**
+ * A customized {@link DoFnRunner} that handles late data dropping and
+ * garbage collect for stateParDo. It registers a GC timer in
+ * processElement and does cleanup in onTimer.
+ *
+ * @param <InputT> the type of the {@link DoFn} (main) input elements
+ * @param <OutputT> the type of the {@link DoFn} (main) output elements
+ */
+public class StatefulDoFnRunner<InputT, OutputT> implements DoFnRunner<InputT, OutputT> {
+
+  public static final String GC_TIMER_ID = "StatefulParDoGcTimerId";
+  public static final String DROPPED_DUE_TO_LATENESS_COUNTER = "StatefulParDoDropped";
+
+  private DoFn<InputT, OutputT> fn;
+  private final SimpleDoFnRunner<InputT, OutputT> doFnRunner;
+  private final StepContext stepContext;
```

Comment:
It is unfortunate to have all of the same parameters passed here that are also provided to the wrapped `DoFnRunner`.

As a first step, I think you could remove the `StepContext` and provide only the `StateInternals` and `TimerInternals`. But going further, you might benefit from  a customized timer setting for cleanup. Something like

```java
class StatefulDoFnRunner {
  interface CleanupTimer {
    void setForWindow(BoundedWindow window);
  }
}
```

For example, a runner might either (a) already know that it always has a timer set for the expiration time or (b) not need a timer at all because it is a batch runner that discards state when it is done. It would be pretty trivial to make a default implementation from `TimerInternals`. This is not a very complete design, but it is just an idea from how I have done cleanup in other contexts.